Initial private SDP interface and naive composite impl

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #81956

Adds an initial private API version of the SDP interface.

Signature:
```
_scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None,
    float dropout_p=0.0, bool need_attn_weights=True, bool is_causal=False) -> (Tensor, Tensor)
```

Returns a tuple of `(output, attn_weights)`.

Note the following:
* `need_attn_weights`: flag indicating that attention weights should be computed. This is useful to toggle off for flash attention as it does not materialize the weights by default, making it more expensive to return them.
* Boolean attention mask support only; `True` values within `attn_mask` indicate that the element should take part in attention (notably, this is reverse of MHA, which uses `True` to mask *out* values). Mask is optional.
* `is_causal`: Temporary flag indicating whether to use a causal attention weighting. If this is set to `True`, it takes precedent over any value passed in for `attn_mask`. Longer term, the `is_causal` flagging can be subsumed into the `attn_mask` arg via tensor subclassing (see e.g. [CausalTensor](https://github.com/facebookresearch/xformers/blob/sparse_cleanup/xformers/sparse/causal_tensor.py) in xFormers).
* Testing is currently done via reference with the existing Python impl of `F._scaled_dot_product_attention`.
* This PR does not yet drop-in the new SDP anywhere. A future PR can hook it up in BT or MHA.

jbschlosser committed:
Initial SDP interface with float masks and naive composite impl

[ghstack-poisoned]

jbschlosser committed:
Update on "Initial SDP interface with float masks and naive composite impl"


Adds an initial version of the SDP interface with:
* A flag `need_attn_weights` for indicating that attention weights do not need to be computed; this is useful for flash attention as it does not materialize the weights, making it more expensive to return them
* Floating point attention mask support (this needs to change, as flash attention only supports causal masks and doesn't require them to be materialized)

[ghstack-poisoned]

jbschlosser committed:
Update on "Initial SDP interface with float masks and naive composite impl"


Adds an initial version of the SDP interface with:
* A flag `need_attn_weights` for indicating that attention weights do not need to be computed; this is useful for flash attention as it does not materialize the weights, making it more expensive to return them
* Floating point attention mask support (this needs to change, as flash attention only supports causal masks and doesn't require them to be materialized)

[ghstack-poisoned]

jbschlosser committed:
Update on "Initial SDP interface with float masks and naive composite impl"


Adds an initial version of the SDP interface with:
* A flag `need_attn_weights` for indicating that attention weights do not need to be computed; this is useful for flash attention as it does not materialize the weights, making it more expensive to return them
* Floating point attention mask support (this needs to change, as flash attention only supports causal masks and doesn't require them to be materialized)

[ghstack-poisoned]

jbschlosser committed:
Update on "Initial SDP interface with float masks and naive composite impl"


Adds an initial private API version of the SDP interface.

Signature:
```
_scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None,
    float dropout_p=0.0, bool need_attn_weights=True, bool is_causal=False) -> (Tensor, Tensor)
```

Returns a tuple of `(output, attn_weights)`.

Note the following:
* `need_attn_weights`: flag indicating that attention weights should be computed. This is useful to toggle off for flash attention as it does not materialize the weights by default, making it more expensive to return them.
* Boolean attention mask support only; `True` values within `attn_mask` indicate that the element should take part in attention (notably, this is reverse of MHA, which uses `True` to mask *out* values). Mask is optional.
* `is_causal`: Temporary flag indicating whether to use a causal attention weighting. If this is set to `True`, it takes precedent over any value passed in for `attn_mask`. Longer term, the `is_causal` flagging can be subsumed into the `attn_mask` arg via tensor subclassing (see e.g. [CausalTensor](https://github.com/facebookresearch/xformers/blob/sparse_cleanup/xformers/sparse/causal_tensor.py) in xFormers).

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -658,5 +658,64 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> native_decoder_only_multi_head_attent
+    // Naive, composite implementation defined here.
+    const auto embed_size = query_.size(-1);
+    const auto query = query_ / ::sqrt(static_cast<double>(embed_size));
+    if (is_causal) {
```
nit: You could throw an exception if "attn_mask" is defined so that the user knows they're about to overwrite it.

jbschlosser replied:
cool, I like this

cpuhrsch commented on test/test_nn.py:
```diff
@@ -20525,6 +20525,49 @@ def test_adaptive_pool_invalid(self, device):
+                      name_fn=lambda dim, is_causal: (f"{dim}D_{'causal_' if is_causal else ''}attn_mask"
+                                                      if dim is not None else "no_attn_mask"))
+    @parametrize_test("dropout_p", [0.0, 0.2, 0.5])
+    def test_scaled_dot_product_attention(self, device, dtype, attn_mask_dim, is_causal, dropout_p):
```
test_nn.py has gotten so long now that Github refuses to do syntax highlighting. Should we move the transformer stuff we're about to add into [test_transformers.py](https://github.com/pytorch/pytorch/blob/master/test/test_transformers.py)?

erichan1 replied:
That would be great (and why test_transformers was made).

jbschlosser replied:
cool, moving it

jbschlosser replied:
FYI it looks like TestTransformers doesn't currently support device-specific tests in the standard way. This can be done easily by:
`instantiate_parametrized_tests(TestTransformers)` -> `instantiate_device_type_tests(TestTransformers, globals())`

Not sure if I want to do it in this PR, will follow-up with this

drisspg commented on test/test_nn.py:
```diff
@@ -20525,6 +20525,49 @@ def test_adaptive_pool_invalid(self, device):
+            if attn_mask_float is not None:
+                attn_mask_float = torch.zeros_like(attn_mask, dtype=query.dtype)
+                attn_mask_float.masked_fill_(attn_mask.logical_not(), float("-inf"))
+            expected = F._scaled_dot_product_attention(
```
Only question I have is that I assume even though this is better transformer specific will ultimately switch out `F. _scaled_dot_product_attention` with the underlying c++ impl. When/If that happens would we just turn this into an op_info?

jbschlosser replied:
We should definitely have an OpInfo for this! I left it as future work because full coverage is non-trivial, but I could be convinced to do it here

drisspg replied:
More curious than anything else what the next steps would be I think this is good

drisspg reviewed (approved):
👍

jbschlosser committed:
Update on "Initial private SDP interface and naive composite impl"


Adds an initial private API version of the SDP interface.

Signature:
```
_scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None,
    float dropout_p=0.0, bool need_attn_weights=True, bool is_causal=False) -> (Tensor, Tensor)
```

Returns a tuple of `(output, attn_weights)`.

Note the following:
* `need_attn_weights`: flag indicating that attention weights should be computed. This is useful to toggle off for flash attention as it does not materialize the weights by default, making it more expensive to return them.
* Boolean attention mask support only; `True` values within `attn_mask` indicate that the element should take part in attention (notably, this is reverse of MHA, which uses `True` to mask *out* values). Mask is optional.
* `is_causal`: Temporary flag indicating whether to use a causal attention weighting. If this is set to `True`, it takes precedent over any value passed in for `attn_mask`. Longer term, the `is_causal` flagging can be subsumed into the `attn_mask` arg via tensor subclassing (see e.g. [CausalTensor](https://github.com/facebookresearch/xformers/blob/sparse_cleanup/xformers/sparse/causal_tensor.py) in xFormers).
* Testing is currently done via reference with the existing Python impl of `F._scaled_dot_product_attention`.
* This PR does not yet drop-in the new SDP anywhere. A future PR can hook it up in BT or MHA.

[ghstack-poisoned]

jbschlosser committed:
Update on "Initial private SDP interface and naive composite impl"


Adds an initial private API version of the SDP interface.

Signature:
```
_scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None,
    float dropout_p=0.0, bool need_attn_weights=True, bool is_causal=False) -> (Tensor, Tensor)
```

Returns a tuple of `(output, attn_weights)`.

Note the following:
* `need_attn_weights`: flag indicating that attention weights should be computed. This is useful to toggle off for flash attention as it does not materialize the weights by default, making it more expensive to return them.
* Boolean attention mask support only; `True` values within `attn_mask` indicate that the element should take part in attention (notably, this is reverse of MHA, which uses `True` to mask *out* values). Mask is optional.
* `is_causal`: Temporary flag indicating whether to use a causal attention weighting. If this is set to `True`, it takes precedent over any value passed in for `attn_mask`. Longer term, the `is_causal` flagging can be subsumed into the `attn_mask` arg via tensor subclassing (see e.g. [CausalTensor](https://github.com/facebookresearch/xformers/blob/sparse_cleanup/xformers/sparse/causal_tensor.py) in xFormers).
* Testing is currently done via reference with the existing Python impl of `F._scaled_dot_product_attention`.
* This PR does not yet drop-in the new SDP anywhere. A future PR can hook it up in BT or MHA.

[ghstack-poisoned]

erichan1 reviewed (approved):
The API looks good to me overall. Is there anything we need to do to check Q,K,V inputs as NestedTensors? ie is any mixture of NestedTensor and regular Tensor fine for Q,K,V? Can we arbitrarily pair NestedTensor Q/K/V and attn_mask and/or is_causal=True?

jbschlosser committed:
Update on "Initial private SDP interface and naive composite impl"


Adds an initial private API version of the SDP interface.

Signature:
```
_scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None,
    float dropout_p=0.0, bool need_attn_weights=True, bool is_causal=False) -> (Tensor, Tensor)
```

Returns a tuple of `(output, attn_weights)`.

Note the following:
* `need_attn_weights`: flag indicating that attention weights should be computed. This is useful to toggle off for flash attention as it does not materialize the weights by default, making it more expensive to return them.
* Boolean attention mask support only; `True` values within `attn_mask` indicate that the element should take part in attention (notably, this is reverse of MHA, which uses `True` to mask *out* values). Mask is optional.
* `is_causal`: Temporary flag indicating whether to use a causal attention weighting. If this is set to `True`, it takes precedent over any value passed in for `attn_mask`. Longer term, the `is_causal` flagging can be subsumed into the `attn_mask` arg via tensor subclassing (see e.g. [CausalTensor](https://github.com/facebookresearch/xformers/blob/sparse_cleanup/xformers/sparse/causal_tensor.py) in xFormers).
* Testing is currently done via reference with the existing Python impl of `F._scaled_dot_product_attention`.
* This PR does not yet drop-in the new SDP anywhere. A future PR can hook it up in BT or MHA.

[ghstack-poisoned]

drisspg commented:
> The API looks good to me overall. Is there anything we need to do to check Q,K,V inputs as NestedTensors? ie is any mixture of NestedTensor and regular Tensor fine for Q,K,V? Can we arbitrarily pair NestedTensor Q/K/V and attn_mask and/or is_causal=True?

I think this is a great callout may want to add some checks to ensure that all QKV have the same "nestedness" status

cpuhrsch commented:
Note that we can always construct a NestedTensor from a regular Tensor. Therefore if *all* inputs can be NestedTensors, so can any mix between a regular Tensor and a NestedTensor.

drisspg commented:
> Note that we can always construct a NestedTensor from a regular Tensor. Therefore if _all_ inputs can be NestedTensors, so can any mix between a regular Tensor and a NestedTensor.

Yes but in the context of this function is it likely that a query would be nested a key non nested and a value nested or any other permutation? Or is it more likely that this is an error in someones code. Do cases like this arise in real transformer code where the query, key, value tensors are generated in different manners such that some are nested and some aren't ?

erichan1 commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -658,5 +658,67 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> native_decoder_only_multi_head_attent
+//     S: Source sequence length
+//     L: Target sequence length
+//     E: Embedding dimension
+std::tuple<Tensor, Tensor> _scaled_dot_product_attention(
```
Adding a comment on code so we can have a thread here. 
- Is there anything we need to do to check Q,K,V inputs as NestedTensors? ie is any mixture of NestedTensor and regular Tensor fine for Q,K,V? 
- Can we arbitrarily pair NestedTensor Q/K/V and attn_mask and/or is_causal=True?

drisspg replied:
Ahh okay +1 on the nit aspect and can be handled when the use case arises. I was more curious than anything but this makes sense.

jbschlosser commented:
> Can we arbitrarily pair NestedTensor Q/K/V and attn_mask and/or is_causal=True?

Setting an explicit `attn_mask` or setting `is_causal=True` are currently *not supported* for Q/K/V nested tensors at the moment due to lacking NT support for:
* empty_like()
* masked_fill_()
* logical_not()
* tril()

When `is_causal=True`, we also need to create an appropriately-sized `attn_mask` with two ragged dims (L, S). Ideally, this would be done without needing specialized logic for nested tensors.

jbschlosser committed:
Update on "Initial private SDP interface and naive composite impl"


Adds an initial private API version of the SDP interface.

Signature:
```
_scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None,
    float dropout_p=0.0, bool need_attn_weights=True, bool is_causal=False) -> (Tensor, Tensor)
```

Returns a tuple of `(output, attn_weights)`.

Note the following:
* `need_attn_weights`: flag indicating that attention weights should be computed. This is useful to toggle off for flash attention as it does not materialize the weights by default, making it more expensive to return them.
* Boolean attention mask support only; `True` values within `attn_mask` indicate that the element should take part in attention (notably, this is reverse of MHA, which uses `True` to mask *out* values). Mask is optional.
* `is_causal`: Temporary flag indicating whether to use a causal attention weighting. If this is set to `True`, it takes precedent over any value passed in for `attn_mask`. Longer term, the `is_causal` flagging can be subsumed into the `attn_mask` arg via tensor subclassing (see e.g. [CausalTensor](https://github.com/facebookresearch/xformers/blob/sparse_cleanup/xformers/sparse/causal_tensor.py) in xFormers).
* Testing is currently done via reference with the existing Python impl of `F._scaled_dot_product_attention`.
* This PR does not yet drop-in the new SDP anywhere. A future PR can hook it up in BT or MHA.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot revert -m "broke all configs on test_scaled_dot_product_attention (__main__.TestNestedTensorAutograd) https://hud.pytorch.org/pytorch/pytorch/commit/f15c5bf13387bde01af48f27d44ad853c004308d" -c landrace

Looks like a landrace but am not sure

jbschlosser commented:
@janeyx99 This is because this PR depends on transpose for nested tensors, and that PR #80981 was reverted.

jbschlosser committed:
Update on "Initial private SDP interface and naive composite impl"


Adds an initial private API version of the SDP interface.

Signature:
```
_scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None,
    float dropout_p=0.0, bool need_attn_weights=True, bool is_causal=False) -> (Tensor, Tensor)
```

Returns a tuple of `(output, attn_weights)`.

Note the following:
* `need_attn_weights`: flag indicating that attention weights should be computed. This is useful to toggle off for flash attention as it does not materialize the weights by default, making it more expensive to return them.
* Boolean attention mask support only; `True` values within `attn_mask` indicate that the element should take part in attention (notably, this is reverse of MHA, which uses `True` to mask *out* values). Mask is optional.
* `is_causal`: Temporary flag indicating whether to use a causal attention weighting. If this is set to `True`, it takes precedent over any value passed in for `attn_mask`. Longer term, the `is_causal` flagging can be subsumed into the `attn_mask` arg via tensor subclassing (see e.g. [CausalTensor](https://github.com/facebookresearch/xformers/blob/sparse_cleanup/xformers/sparse/causal_tensor.py) in xFormers).
* Testing is currently done via reference with the existing Python impl of `F._scaled_dot_product_attention`.
* This PR does not yet drop-in the new SDP anywhere. A future PR can hook it up in BT or MHA.

[ghstack-poisoned]

jbschlosser committed:
Update on "Initial private SDP interface and naive composite impl"


Adds an initial private API version of the SDP interface.

Signature:
```
_scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None,
    float dropout_p=0.0, bool need_attn_weights=True, bool is_causal=False) -> (Tensor, Tensor)
```

Returns a tuple of `(output, attn_weights)`.

Note the following:
* `need_attn_weights`: flag indicating that attention weights should be computed. This is useful to toggle off for flash attention as it does not materialize the weights by default, making it more expensive to return them.
* Boolean attention mask support only; `True` values within `attn_mask` indicate that the element should take part in attention (notably, this is reverse of MHA, which uses `True` to mask *out* values). Mask is optional.
* `is_causal`: Temporary flag indicating whether to use a causal attention weighting. If this is set to `True`, it takes precedent over any value passed in for `attn_mask`. Longer term, the `is_causal` flagging can be subsumed into the `attn_mask` arg via tensor subclassing (see e.g. [CausalTensor](https://github.com/facebookresearch/xformers/blob/sparse_cleanup/xformers/sparse/causal_tensor.py) in xFormers).
* Testing is currently done via reference with the existing Python impl of `F._scaled_dot_product_attention`.
* This PR does not yet drop-in the new SDP anywhere. A future PR can hook it up in BT or MHA.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
